File: 08-Singleton/src/go/database_connection.go
package main

import (
	"fmt"
	"sync"
)

type DatabaseConnection struct {
	host      string
	port      int
	connected bool
}

var (
	dbInstance *DatabaseConnection
	dbOnce     sync.Once
)

// GetDatabaseInstance returns the singleton DatabaseConnection.
// sync.Once guarantees the init function runs exactly once,
// even with concurrent goroutines.
func GetDatabaseInstance() *DatabaseConnection {
	dbOnce.Do(func() {
		dbInstance = &DatabaseConnection{
			host:      "localhost",
			port:      5432,
			connected: false,
		}
	})
	return dbInstance
}

func (db *DatabaseConnection) Connect() {
	if !db.connected {
		db.connected = true
		fmt.Printf("Connected to %s:%d\n", db.host, db.port)
	} else {
		fmt.Printf("Already connected to %s:%d\n", db.host, db.port)
	}
}

func (db *DatabaseConnection) Disconnect() {
	if db.connected {
		db.connected = false
		fmt.Printf("Disconnected from %s:%d\n", db.host, db.port)
	}
}

func (db *DatabaseConnection) Query(sql string) string {
	if !db.connected {
		panic("Not connected — call Connect() first")
	}
	return "Result of: " + sql
}

func (db *DatabaseConnection) IsConnected() bool {
	return db.connected
}

func (db DatabaseConnection) String() string {
	return fmt.Sprintf("DatabaseConnection[%s:%d, connected=%t]", db.host, db.port, db.connected)
}

File: 08-Singleton/src/rust/database_connection.rs
use std::collections::HashMap;
use std::sync::OnceLock;

pub struct DatabaseConnection {
    host: String,
    port: u16,
    connected: bool,
}

// Global singleton using OnceLock (stable since Rust 1.80).
// OnceLock guarantees thread-safe, lazy, one-time initialization.
// We use a Mutex to allow interior mutability (connect/disconnect).
static DB_INSTANCE: OnceLock<std::sync::Mutex<DatabaseConnection>> = OnceLock::new();

impl DatabaseConnection {
    fn new() -> Self {
        DatabaseConnection {
            host: String::from("localhost"),
            port: 5432,
            connected: false,
        }
    }

    pub fn get_instance() -> &'static std::sync::Mutex<DatabaseConnection> {
        DB_INSTANCE.get_or_init(|| std::sync::Mutex::new(DatabaseConnection::new()))
    }

    pub fn connect(&mut self) {
        if !self.connected {
            self.connected = true;
            println!("Connected to {}:{}", self.host, self.port);
        } else {
            println!("Already connected to {}:{}", self.host, self.port);
        }
    }

    pub fn disconnect(&mut self) {
        if self.connected {
            self.connected = false;
            println!("Disconnected from {}:{}", self.host, self.port);
        }
    }

    pub fn query(&self, sql: &str) -> String {
        if !self.connected {
            panic!("Not connected — call connect() first");
        }
        format!("Result of: {}", sql)
    }

    pub fn is_connected(&self) -> bool {
        self.connected
    }
}

impl std::fmt::Display for DatabaseConnection {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "DatabaseConnection[{}:{}, connected={}]",
            self.host, self.port, self.connected
        )
    }
}
